# CAP-5510 Term Project

## Project Title

Implementation and Analysis of different Algorithms to find Tandem Repeats in
DNA sequences.

## Team members 

- Tushar Kaley (9126-1421)
- Akash Shingte (4874-1966)

## Abstract

A tandem repeat in DNA is two or more contiguous, approximate copies of a pattern of nucleotides. Tandem repeats have been shown to cause human disease, may play a variety of regulatory and evolutionary roles and are important laboratory and analytic tools.

Extensive knowledge about pattern size, copy number, mutational history, etc. for tandem repeats has been limited by the inability to easily detect them in genomic sequence data. For this purpose, a variety of algorithms exists for finding tandem repeats. 

In this project we propose to implement, analyze and compare different algorithms for finding tandem repeats. Namely, 
1. Suffix Array Implementation
2. Suffix Tree Implementation
3. Dynamic Programming Implementation

## Plan of action
1. Read and understand the algorithms papers
2. Implementation of suffix array algorithm
3. Implementation of suffix tree algorithm
4. Implementation of dynamic programming algorithm
5. Comparison of all algorithms.

## Distribution of workload
The workload will be distributed evenly across both team members. A rough
distribution of workload - 

1. Suffix Tree (Akash)
2. Suffix Array Implementation (Tushar)
3. Dynamic Programming (Akash & Tushar)
4. Comparison (Akash & Tushar)

## References

1. Simple and Flexible Detection of Contiguous Repeats Using a Suffix Tree  - by Jens Stoye and Dan Gusfield
2. Space Efficient Linear Time construction of Suffix Arrays - by Pang Ko and Srinivas Aluru
3. Finding Repeats within Strings - by Dina Sokol, Fredrick Adkins, Zhongyuan Che, Kristin Pfabe
4. SA-SSR: A Suffix Array-Based Algorithm for Exhaustive and Efficient SSR Discovery in Large Genetic Sequences - by Brandon Pickett, S.M. Karlinsey, C.E. Penrod, Perry G Ridge
5. http://dimacs.rutgers.edu/Publications/Modules/Module09-2/dimacs09-2.pdf - Dynamic programming based Algorithm
6. http://www.sciencedirect.com/science/article/pii/S1570866704000498 - Suffix Array based Algorithm
7. http://www.lirmm.fr/~rivals/ALGOSEQ/DOC/TREvolSurvey.pdf
8. Benson, Gary. “Tandem repeats finder: a program to analyze DNA sequences.” Nucleic acids research 27 2 (1999): 573-80. [Link](https://www.semanticscholar.org/paper/Tandem-repeats-finder%3A-a-program-to-analyze-DNA-Benson/991a7744b15453096e756a2d8ea0e44ee30738ae)
9. Olivier Delgrange, Eric Rivals. "STAR: An Algorithm to Search for Tandem Approximate Repeats". Bioinformatics, Oxford University Press (OUP), 2004, 20 (16), pp.2812-2820 [Link](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.860.1856&rep=rep1&type=pdf)
